### PR TITLE
fix(rust, python): raise on unsupported transpose and object types

### LINKED ISSUE #9927: The shortest way to segfault polars via `DataFrame.transpose()`
### Checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
>>> pl.DataFrame([object()]).transpose()
memory allocation of 4539145048 bytes failed

>>> pl.DataFrame({'x': object()}).transpose()
memory allocation of 4874689368 bytes failed
```

### Installed versions

<details>

```
--------Version info---------
Polars:              0.18.7
Index type:          UInt32
Platform:            macOS-13.4.1-arm64-arm-64bit
Python:              3.11.4 | packaged by conda-forge | (main, Jun 10 2023, 18:08:41) [Clang 15.0.7 ]

----Optional dependencies----
adbc_driver_sqlite:  <not installed>
connectorx:          <not installed>
deltalake:           <not installed>
fsspec:              <not installed>
matplotlib:          3.7.2
numpy:               1.24.4
pandas:              2.0.3
pyarrow:             11.0.0
pydantic:            <not installed>
sqlalchemy:          <not installed>
xlsx2csv:            <not installed>
xlsxwriter:          <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by ritchie46**
We don't have the architecture in place yet to support objects. Let's raise for now.

fixes #9927

### CODE REVIEW

**Comment by ritchie46** [comment]
Those errors are really strange? They seem totally unrelated and are influence by how many tests we run. E.g. running all tests, some fail (only one locally on my side). Running that test by itself, it passes.

**Comment by ritchie46** [comment]
@alexander-beedie any idea what happens here?

```python
polars/lazyframe/frame.py:295: RuntimeError
___________________________________ test_init_structured_objects ____________________________________

    def test_init_structured_objects() -> None:
        # validate init from dataclass, namedtuple, and pydantic model objects
        from typing import NamedTuple
    
        from polars.dependencies import dataclasses, pydantic
    
        @dataclasses.dataclass
        class TeaShipmentDC:
            exporter: str
            importer: str
            product: str
            tonnes: int | None
    
        class TeaShipmentNT(NamedTuple):
            exporter: str
            importer: str
            product: str
            tonnes: None | int
    
        class TeaShipmentPD(pydantic.BaseModel):
            exporter: str
            importer: str
            product: str
            tonnes: int
    
        for Tea in (TeaShipmentDC, TeaShipmentNT, TeaShipmentPD):
            t0 = Tea(exporter="Sri Lanka", importer="USA", product="Ceylon", tonnes=10)
            t1 = Tea(exporter="India", importer="UK", product="Darjeeling", tonnes=25)
            t2 = Tea(exporter="China", importer="UK", product="Keemum", tonnes=40)
    
>           s = pl.Series("t", [t0, t1, t2])

tests/unit/series/test_series.py:180: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
polars/series/series.py:276: in __init__
    self._s = sequence_to_pyseries(
polars/utils/_construction.py:379: in sequence_to_pyseries
    return pl.DataFrame(values).to_struct(name)._s
polars/dataframe/frame.py:361: in __init__
    self._df = sequence_to_pydf(
polars/utils/_construction.py:846: in sequence_to_pydf
    return _sequence_to_pydf_dispatcher(
../../../miniconda3/lib/python3.10/functools.py:889: in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
polars/utils/_construction.py:1088: in _sequence_of_elements_to_pydf
    pl.Series(column_names[0], data, schema_overrides.get(column_names[0]))._s
polars/series/series.py:276: in __init__
    self._s = sequence_to_pyseries(
polars/utils/_construction.py:379: in sequence_to_pyseries
    return pl.DataFrame(values).to_struct(name)._s
polars/dataframe/frame.py:361: in __init__
    self._df = sequence_to_pydf(
polars/series/series.py:276: in __init__
    self._s = sequence_to_pyseries(
polars/utils/_construction.py:379: in sequence_to_pyseries
    return pl.DataFrame(values).to_struct(name)._s
polars/dataframe/frame.py:361: in __init__
    self._df = sequence_to_pydf(
polars/utils/_construction.py:846: in sequence_to_pydf
    return _sequence_to_pydf_dispatcher(
../../../miniconda3/lib/python3.10/functools.py:889: in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
polars/utils/_construction.py:1088: in _sequence_of_elements_to_pydf
    pl.Series(column_names[0], data, schema_overrides.get(column_names[0]))._s
polars/series/series.py:275: in __init__
    elif isinstance(values, Sequence):
../../../miniconda3/lib/python3.10/typing.py:994: in __instancecheck__
    return self.__subclasscheck__(type(obj))
../../../miniconda3/lib/python3.10/typing.py:1158: in __subclasscheck__
    return issubclass(cls, self.__origin__)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cls = <class 'collections.abc.Sequence'>, subclass = <class 'list'>

    def __subclasscheck__(cls, subclass):
        """Override for issubclass(subclass, cls)."""
>       return _abc_subclasscheck(cls, subclass)
E       RecursionError: maximum recursion depth exceeded in comparison

../../../miniconda3/lib/python3.10/abc.py:123: RecursionError
!!! Recursion error detected, but an error occurred locating the origin of recursion.
  The following exception happened when comparing locals in the stack frame:
    AttributeError: 'NoneType' object has no attribute 'dtype'
  Displaying first and last 10 stack frames out of 719.

```

File: py-polars/tests/unit/operations/test_transpose.py [hunk 1]
```diff
@@ -147,3 +147,8 @@ def test_transpose_logical_data() -> None:
         }
     )
     assert_frame_equal(result, expected)
+
+
+def test_err_transpose_object() -> None:
```
Comment (ritchie46) [inline py-polars/tests/unit/operations/test_transpose.py:152]: Somehow, removing this test resolves it.

**Comment by ritchie46** [comment]
@alexander-beedie nvm, got a workaround. There is definitely something strange going on. 

This was helped by replacing the `object` type with a (less deep?) `CustomObject`?

```python
   with pytest.raises(pl.InvalidOperationError):
        pl.DataFrame([object()]).transpose()
```

```python
    class CustomObject:
        pass

    with pytest.raises(pl.InvalidOperationError):
        pl.DataFrame([CustomObject()]).transpose()
```

Still strange, but no blocker. :)

**Comment by alexander-beedie** [comment]
That looks... very weird indeed! No clue what's happening without a repro 😅

**Pull request merged**